### PR TITLE
fix: Migrate remaining components to TypeScript

### PR DESCRIPTION
**Pull request by markdalgleish**
This PR migrates the remaining JavaScript components to TypeScript:
- Checkbox
- ChecklistCard
- FieldMessage
- Radio

This also performs a number of refactors and clean-ups:
- Deleted the `withTheme` HOC in favour of `ThemeConsumer`, since render props are _dramatically_ easier to type.
- Exported all component prop interfaces to allow other components to easily extend them.
- Consistently named prop interfaces, e.g. a `Foo` component should have a `FooProps` interface, rather than just `Props`. This is partly to ensure that error messages from TypeScript are more readable when interfaces are not satisfied.
- Ensured all components extend prop interfaces correctly when spreading props, e.g. `export interface TextProps extends BoxProps { ... }`
- Removed props from interfaces that are duplicates of inherited props, e.g. `className`.
- Generally expanded the type annotations in existing TypeScript files (not merely to satisfy the compiler, but to better document the code).
- Removed `react-docgen` since we no longer use `prop-types` in Braid components.
- Moved `prop-types` to `devDependencies` (until the documentation site is also migrated to TypeScript).

### CODE REVIEW

File: lib/components/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+import TickIcon from '../icons/TickIcon/TickIcon';
+import styles from './Checkbox.css.js';
+import { px } from '../../themes/utils/toUnit';
+
+const textColorForState = ({
+  disabled,
+  hovered
+}: {
+  disabled: boolean | undefined;
+  hovered: boolean | undefined;
```
Comment (mattsjones) [inline lib/components/Checkbox/Checkbox.tsx]: Think this should just be `boolean` as the state inits with `false`.

File: lib/components/Checkbox/Checkbox.tsx [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+  render() {
+    return (
+      <ThemeConsumer>
+        {theme => {
+          const {
+            variant,
+            id,
+            checked,
+            label,
+            disabled,
```
Comment (mattsjones) [inline lib/components/Checkbox/Checkbox.tsx]: Thoughts on making `disabled` have a default value of `false`? I think that'd be cleaner.

File: lib/components/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                      paddingBottom="small"
+                      className={styles.children}
+                      style={{ marginLeft: px(checkboxSize) }}
+                    >
+                      {children}
+                    </Box>
+                  </Box>
+                ) : null}
+              </div>
+              {message !== false ? (
```
Comment (mattsjones) [inline lib/components/Checkbox/Checkbox.tsx]: Is this a tri-state? e.g. `true`, `false`, `undefined`?
If so can we represent that as an enum instead?

File: lib/components/Radio/Radio.tsx [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+import FieldMessage from '../FieldMessage/FieldMessage';
+import styles from './Radio.css.js';
+import { px } from '../../themes/utils/toUnit';
+
+const textColorForState = ({
+  disabled,
+  hovered
+}: {
+  disabled: boolean | undefined;
+  hovered: boolean | undefined;
```
Comment (mattsjones) [inline lib/components/Radio/Radio.tsx]: Same as above

File: lib/components/Text/Text.tsx [hunk 5]
```diff
@@ -26,48 +24,51 @@ const isTransformVariant = (
...
+          const {
+            component,
+            size,
+            color,
+            weight,
+            baseline = true,
+            className,
+            ...restProps
+          } = this.props;
+          const fontSize = size || 'standard';
```
Comment (mattsjones) [inline lib/components/Text/Text.tsx]: Can we just make `size` has a default value of `standard`?

File: lib/components/icons/Icon/Icon.tsx [hunk 6]
```diff
@@ -21,44 +18,47 @@ const isSizeVariant = (
...
-        />
-      );
-    }
+            size,
+            svgComponent,
+            inline = false,
+            fill,
+            ...restProps
+          } = this.props;
+          const sizeAtom = `${size || 'standard'}Text${inline ? 'Inline' : ''}`;
```
Comment (mattsjones) [inline lib/components/icons/Icon/Icon.tsx]: Can size have a default value of `standard`?

File: lib/components/Checkbox/Checkbox.tsx (hunk 3 above)
Comment (markdalgleish) [reply lib/components/Checkbox/Checkbox.tsx]: Nope, it's either a string, or `false` to disable the message entirely.

**Comment by markdalgleish** [comment]
@mattsjones Responded to all feedback, just pushed an update.

**Pull request merged**